### PR TITLE
[NTOS:KE/x64] Improve exception handling

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Improve handling of exceptions.

## Proposed changes

- [NTOS:KE/x64] Do not overwrite the original context during exception handling
- [NTOS:KE/x64] Fix exception information on page faults

## Tests
- ✅ KVM x64: https://reactos.org/testman/compare.php?ids=97756,97760,97774,97782

### CODE REVIEW

File: ntoskrnl/ke/amd64/trap.S [hunk 1]
```diff
@@ -461,8 +461,11 @@ IntsDisabled:
 
 PageFaultError:
 
-    /* Set parameter 1 to error code */
+    /* Set parameter 1 to write/execute flag.
+       See https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-exception_record */
     mov r9d, [rbp + KTRAP_FRAME_ErrorCode]
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/trap.S:466]: Can you tell me in comment what the format of this error code initially stored in r9d is? (and how we come up with the correct flags you want afterwards with the bit flagging done below)

**Review by HBelusca — COMMENTED** [review]
Lgtm; just a minor comment and a native question.

**Review by HBelusca — APPROVED** [review]

File: sdk/lib/rtl/amd64/unwind.c [hunk 2]
```diff
@@ -708,8 +708,11 @@ RtlpUnwindInternal(
             UnwindContext.Rip = *(DWORD64*)UnwindContext.Rsp;
             UnwindContext.Rsp += sizeof(DWORD64);
 
-            /* Copy the context back for the next iteration */
-            *ContextRecord = UnwindContext;
+            if (HandlerType == UNW_FLAG_UHANDLER)
+            {
+                /* Copy the context back for the next iteration */
+                *ContextRecord = UnwindContext;
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply sdk/lib/rtl/amd64/unwind.c:714]: True. But this function is subject to be refactored anyway. I'm just trying to fix some of the easier to do things first.

File: sdk/lib/rtl/amd64/unwind.c [hunk 3]
```diff
@@ -756,7 +759,7 @@ RtlpUnwindInternal(
 
             /* Log the exception if it's enabled */
             RtlpCheckLogException(ExceptionRecord,
-                                  ContextRecord,
+                                  &UnwindContext,
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply sdk/lib/rtl/amd64/unwind.c:762]: No, it should be correct as it is. There is likely still lots of things wrong, but it's better than before.

File: ntoskrnl/ke/amd64/trap.S (hunk 1 above)
Comment (tkreuzer) [reply ntoskrnl/ke/amd64/trap.S:466]: The CPU pushes the fault code: http://wiki.osdev.org/Exceptions#Page_Fault
0x02 (bit 1) = Write, 0x10 (bit 4) = Instruction Fetch, mask them and shift them right by 1 and you get the flags for the exception.

**Pull request merged**